### PR TITLE
[WIP] Fixed #34077 -- Made BoundField renderable.

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket 43077](https://code.djangoproject.com/ticket/34077)

### CODE REVIEW

**Review by carltongibson — COMMENTED** [review]
Quick initial comment: this should have a major section at the top of the release notes, with a full usage example, I think. (We wanna tell folks about it! 😜)

Nice work picking this up David 🦄

File: django/forms/boundfield.py [hunk 1]
```diff
@@ -81,6 +82,15 @@ def errors(self):
...
         )
 
+    def as_field(self, template_name=None, context=None, renderer=None):
+        """
+        Render the field including labels, help text and errors.
+        """
+        renderer = renderer or self.form.renderer
+        template = template_name or self.field.template_name
+        context = context or {"field": self}
+        return mark_safe(renderer.render(template, context))
```
[the remark this answers is not in the document]
Comment (smithdc1) [reply django/forms/boundfield.py]: This highlighted a nice bug I'd likely introduced. Hopefully we'll have one test failure showing this. 

We have a circular series of renderables as the form contains its boundfield and the boundfield has a reference back to the form. 

While I need to spend more time looking at this I'd also appreciate ideas.

**Comment by carltongibson** [comment]
Hey @smithdc1 — Just looking at the discussion here. Can I ask you to ping me when you're ready for a review. Thanks. 🎁

**Comment by smithdc1** [comment]
@carltongibson -- Can I ask you to put this on your list please? 

On the circular rendering issue, my current thinking is that it's actually ok _for now_. The original issue was recurively calling `__str__` gave rise to a circular context. This patch doesn't change `BoundField.__str__` and therefore I don't belive this patch as is should be causing that regession. I think the error is that the test incorrectly calls `render()`. 

I've changed the test to call `__str__()` instead. I believe this to still be a regression test for https://code.djangoproject.com/ticket/33134 but also allows the tests to pass with this patch. 🤞 

The _for now_ reference is that if we ever did want to change `BoundField.__str__()` to render the field then we'd encounter that regression test.

**Review by ngnpope — COMMENTED** [review]
Thanks David. This is a good next step, but I think we need to think about extracting the field rendering for more than just the `div.html`. Really `.as_field()` should be `.as_div()` and we should add the equivalent for the other tags.

I'll try and have a think through the mess with `.as_widget()` and `.as_hidden()` in `.__str__()`. It feels pretty clunky and looks sort of outdated w.r.t. moving stuff to template rendering. Biggest pain is that they're documented API.

**Comment by smithdc1** [comment]
Thank you for you review and comments as always. 

Just a couple of initial comments. 

> we should add the equivalent for the other tags.

We don't recommend the other methods. We stopped short of removing them (that was the suggestion by the accessibility team) due to backward compatibility concerns. So I wasn't sure on adding additional support for them. 

Maybe more importantly I'm not sure there's that much benefit. How many folk are using _exactly_ the standard layout for those methods and would benefit (e.g im not sure how altering the table layout would help)? As soon as you have something slightly different the advise would be to create a custom template and use that instead? Assign it in the renderer or field or ...

> Biggest pain is that they're documented API.

It's also been like this for _some time_. 
https://github.com/django/django/blame/4d596a1f6443eaf5d18d70a18aaac25030c7fc81/django/newforms/__init__.py#L829

I had assumed that was _too much_ to be changing at this stage.

File: django/forms/utils.py [hunk 2]
```diff
@@ -78,6 +78,31 @@ def render(self, template_name=None, context=None, renderer=None):
...
+    def as_widget(self):
+        raise NotImplementedError(
+            "Subclasses of RenderableFieldMixin must provide an as_widget() method."
+        )
+
+    def __str__(self):
+        """Render this field as an HTML widget."""
+        if self.field.show_hidden_initial:
+            return self.as_widget() + self.as_hidden(only_initial=True)
+        return self.as_widget()
```
[the remark this answers is not in the document]
Comment (smithdc1) [reply django/forms/utils.py:101]: This mornings thought was I wonder if the first step here would be to see if we can make `Widget` inherit from `RenderableMixin` or similar. 

I don't think we can currently call `str()` on a widget to get it to render. That could be nice in a template. 🤔 

```
{{ field.label }}
{{ field.help_text }}
{{ field.widget }}
...
```

Thinking out loud here!

**Comment by carltongibson** [comment]
> We don't recommend the other methods.

Yes... — if we were starting from scratch now, we wouldn't have the `as_*()` methods on form: rather you'd just set the template and output as a string (`str(form)`) 

So here… — if we were starting from scratch now, we'd do the same for `BoundField`, except we can't because `__str__()` is already taken. Thus, we need to add the new `as_field()` (still thinking about the name there) in order to allow folks to output the errors/help-text/widget group all in one go. (The docs should be clear that `str(bf)` just output the widget, where `bf.as_field()` outputs the collected parts.) 

I don't think we need (or, for the a11y concerns, want) to add `as_p()` &co like templates. Let's just provide one that. The div templates are becoming the default, let's go with that. (Any issues there? 🤔) 


> On the circular rendering issue... 

👀

**Comment by ngnpope** [comment]
> Yes... — if we were starting from scratch now, we wouldn't have the as_*() methods on form: rather you'd just set the template and output as a string (str(form))

Ah, if only, eh? 🙂

> I don't think we need (or, for the a11y concerns, want) to add as_p() &co like templates. Let's just provide one that. The div templates are becoming the default, let's go with that. (Any issues there? thinking)

If we're to do this - only focus on improving the `.as_div()` case - I would like to see a stronger admonition under the docs for `.as_p()`, `.as_table()`, and `.as_ul()`. It should state that we're no longer developing improvements around those helpers, that they're poor for accessibility reasons, and that we recommend migrating over to using `.as_div()` instead. We should possibly even say that they're soft-deprecated. Currently we only [mention this](https://docs.djangoproject.com/en/4.1/ref/forms/api/#django.forms.Form.as_div) under `.as_div()`, but it doesn't feel strong enough (to me) to warn people away from them.

Not splitting out those cases and advertising this split-out field template as a new feature may be confusing to users of those non-recommended methods.

**Comment by carltongibson** [comment]
Hey @ngnpope. 👍 

We already moved the `as_*` methods to the bottom of _Outputting forms as HTML_, in the subsection here: https://docs.djangoproject.com/en/4.1/ref/forms/api/#output-styles, that you linked. This was an attempt to de-emphasise them. We could move the note to the top of that subsection, and add another sentence saying, ≈_"In general the `{{ form }}` usage is preferred. These helpers largely remain for backwards compatibility reasons."_

**Comment by smithdc1** [comment]
I'll come back to this after the feature freeze. 🎅🎄

I've spent a bit of time looking at the comments and it seems to unravel and get more complex the more I think through it. I'm not sure unpicking the current str behaviour is that straight forward. 🤔

**Comment by carltongibson** [comment]
> I'll come back to this after the feature freeze. 🎅🎄

Hey @smithdc1 — OK. Always wise 💝 — Overall, I think this is a (really) good feature to have, but there's no urgency. 5.0 will be more than perfect. 🎁

**Comment by smithdc1** [comment]
Hey @carltongibson 👋

Just to say, I'm planning on coming back to this next. I've got one eye on trying to get this in before you move onto new adventures. 😀

**Comment by carltongibson** [comment]
Cool. Yes. Let's get it in. It's the icing on the cake for the forms changes over the last few releases. 

(I didn't get to think about it but, did we consider the template tag approach? Do we favour the method on BoundField?)

**Comment by smithdc1** [comment]
> (I didn't get to think about it but, did we consider the template tag approach? Do we favour the method on BoundField?)

I think I originally suggested this as an option on the ticket. I've come round to the idea that it is not the best option here. All of the other ways of rendering a form/field/widget are a method, usually one of `__str__()`, `render()` or one of the `as_*()` methods.

To suddenly have a template tag to render a field seems like an odd choice. 

I think the most difficult question we have to answer is what do we do about calling `__str__` on a `BoundField`, and the fact that it returns the fields widget. As Nick pointed out it's not just `__str__` there's also `.as_widget()` and `.as_hidden()` so it's quite a lot to try and unpick.  It's also been that way ~forever. Do we try and deprecate that so we can clean up the rest of the code?

> We could move the note in the as_div() bit to the top of that subsection ...

I'll think I'll start here, likely as another PR so we can get that in.

File: docs/releases/5.0.txt [hunk 3]
```diff
@@ -37,6 +37,70 @@ compatible with Django 5.0.
 What's new in Django 5.0
 ========================
 
+BoundField becomes renderable
```
Comment (smithdc1) [inline docs/releases/5.0.txt:40]: This needs a better title.

**Comment by carltongibson** [comment]
Closing in favour of #16574

**Pull request closed without merging**